### PR TITLE
Fix `cupy.fill` to properly take zero-dim `cupy.ndarray`

### LINKED ISSUE #6437: `cupy.fill` fails to take a zero-dim cupy.ndarray with a different dtype
Found in #6181. `cupy.fill` does not accept a zero-dim cupy.ndarray with a different dtype and raises a TypeError.

```py
>>> import cupy
>>> a = cupy.empty((1,), dtype=cupy.int64)
>>> a.fill(cupy.array(1, dtype=cupy.int32))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "cupy/_core/core.pyx", line 665, in cupy._core.core.ndarray.fill
    cpdef fill(self, value):
  File "cupy/_core/core.pyx", line 683, in cupy._core.core.ndarray.fill
    fill_kernel(value, self)
  File "cupy/_core/_kernel.pyx", line 839, in cupy._core._kernel.ElementwiseKernel.__call__
    in_types, out_types, type_map = self._decide_params_type(
  File "cupy/_core/_kernel.pyx", line 882, in cupy._core._kernel.ElementwiseKernel._decide_params_type
    ret = _decide_params_type_core(
  File "cupy/_core/_kernel.pyx", line 536, in cupy._core._kernel._decide_params_type_core
    raise TypeError(
TypeError: Type is mismatched. x <class 'numpy.int32'> <class 'numpy.int64'> T
```

```py
>>> import numpy
>>> a = numpy.empty((1,), dtype=numpy.int64)
>>> a.fill(numpy.array(1, dtype=numpy.int32))
>>> a
array([1])
```

### PR DESCRIPTION
**Pull request by takagi**
Close #6437.

### CODE REVIEW

File: tests/cupy_tests/core_tests/test_ndarray_copy_and_view.py [hunk 1]
```diff
@@ -143,10 +143,21 @@ def test_fill_with_numpy_scalar_ndarray(self, xp, dtype):
         return a
 
     @testing.for_all_dtypes()
-    def test_fill_with_numpy_nonscalar_ndarray(self, dtype):
+    def test_fill_with_cupy_scalar_ndarray(self, dtype):
```
Comment (toslunar) [inline tests/cupy_tests/core_tests/test_ndarray_copy_and_view.py:146]: How about using `@testing.numpy_cupy_array_equal`?

File: tests/cupy_tests/core_tests/test_ndarray_copy_and_view.py [hunk 2]
```diff
@@ -143,10 +143,21 @@ def test_fill_with_numpy_scalar_ndarray(self, xp, dtype):
         return a
 
     @testing.for_all_dtypes()
```
Comment (toslunar) [inline tests/cupy_tests/core_tests/test_ndarray_copy_and_view.py:145]: How about testing casts by `for_all_dtypes_combination`?

File: cupy/_core/core.pyx [hunk 3]
```diff
@@ -671,6 +671,14 @@ cdef class ndarray:
         .. seealso:: :meth:`numpy.ndarray.fill`
 
         """
+        if isinstance(value, cupy.ndarray):
+            if value.shape != ():
+                raise ValueError(
+                    'non-scalar cupy.ndarray cannot be used for fill')
```
Comment (toslunar) [inline cupy/_core/core.pyx:677]: The message LGTM because it's consistent with the `numpy.ndarray` case below.

**Review by toslunar — CHANGES_REQUESTED** [review]
The implementation LGTM and I roughly checked that both NumPy and CuPy `fill` the value like `copyto`, i.e. the `'unsafe'` casting with ComplexWarning.

I left comments on the test.  Besides, it seems better to compare the 3 cases

- `numpy.ndarray.fill(numpy.ndarray)`
- `cupy.ndarray.fill(numpy.ndarray)`
- `cupy.ndarray.fill(cupy.ndarray)`

but such refactoring is optional.  (I don't come up with an elegant way.)

**Comment by takagi** [comment]
PTAL! For comparing the 3 cases above, I think comparing the following two makes sense:
- `numpy.ndarray.fill(numpy.ndarray)` and `cupy.ndarray.fill(numpy.ndarray)`
- `numpy.ndarray.fill(numpy.ndarray)` and `cupy.ndarray.fill(cupy.ndarray)`

They are corresponding to `test_fill_with_numpy_scalar_ndarray` and `test_fill_with_cupy_scalar_ndarray`.

**Review by toslunar — APPROVED** [review]
LGTM

**Pull request merged**